src/hosting/runTags.ts

```diff
@@ -0,0 +1,183 @@
+ */
+export async function setRewriteTags(
+  rewrites: api.Rewrite[],
+  project: string,
+  version: string
+): Promise<void> {
+  // Note: this is sub-optimal in the case where there are multiple rewrites
+  // to the same service. Should we deduplicate this?
+  const services: run.Service[] = await Promise.all(
+    rewrites
+      .map((rewrite) => {
+        if (!("run" in rewrite)) {
+          return null;
+        }
+        if (rewrite.run.tag !== TODO_TAG_NAME) {
+          return null;
+        }
+
+        return run.getService(
+          `projects/${project}/locations/${rewrite.run.region}/services/${rewrite.run.serviceId}`
+        );
+      })
+      // filter does not drop the null annotation
+      .filter((s) => s !== null) as Array<Promise<run.Service>>
+  );
+  // Unnecessary due to functional programming, but creates an observable side effect for tests
+  if (!services.length) {
+    return;
+  }
+
+  const needsGC = services
+    .map((service) => {
+      return service.spec.traffic.filter((traffic) => traffic.tag).length;
+    })
+    .some((length) => length >= garbageCollectionThreshold);
+  if (needsGC) {
+    await exports.gcTagsForServices(project, services);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+  const tags: Record<string, Record<string, string>> = await exports.ensureLatestRevisionTagged(
+    services,
+    `fh-${version}`
+  );
+  for (const rewrite of rewrites) {
+    if (!("run" in rewrite) || rewrite.run.tag !== TODO_TAG_NAME) {
+      continue;
+    }
+    const tag = tags[rewrite.run.region][rewrite.run.serviceId];
+    rewrite.run.tag = tag;
+  }
+}
+
+/**
+ * Given an already fetched service, ensures that the latest revision
+ * has a tagged traffic target.
+ * If the service does not have a tagged target already, the service will be modified
+ * to include a new target and the change will be publisehd to prod.
+ * Returns a map of region to map of service to latest tag.
+ */
+export async function ensureLatestRevisionTagged(
+  services: run.Service[],
+  defaultTag: string
+): Promise<Record<string, Record<string, string>>> {
+  // Region -> Service -> Tag
+  const tags: Record<string, Record<string, string>> = {};
+  const updateServices: Array<Promise<unknown>> = [];
+  for (const service of services) {
+    const { projectNumber, region, serviceId } = run.gcpIds(service);
+    tags[region] = tags[region] || {};
+    const latestRevisionTarget = service.status?.traffic.find((target) => target.latestRevision);
+    if (!latestRevisionTarget) {
+      throw new FirebaseError(
+        `Assertion failed: service ${service.metadata.name} has no latestRevision traffic target`
+      );
+    }
+    const latestRevision = latestRevisionTarget.revisionName;
+    const alreadyTagged = service.spec.traffic.find(
+      (target) => target.revisionName === latestRevision && target.tag
+    );
+    if (alreadyTagged) {
+      tags[region][serviceId] = alreadyTagged.tag!;
```

Comment:
It's fairly obvious, but put a comment here that the null assertion is okay 'cause of the above check, please :)